pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -618,7 +618,11 @@ func updateStatefulSetAfterInvariantEstablished(
 		if err != nil {
 			return &status, err
 		}
-
+		// maxUnavailable might be zero for small percentage without round up.
```

Comment:
If `maxUnavailable` can never be less than 1, maybe this can be fixed by adding a validation? WDYT?
https://github.com/kubernetes/kubernetes/blob/f85ff4b5743d501381c76b94a5bc6197b7766190/pkg/apis/apps/validation/validation.go#L93

Reply:
Actually I thought about this, but it is not a good idea. The validation only has to make sure `maxUnavailable > 0` if it is an integer or `0% < maxUnavailable <= 100%` if it is a percentage.

But if we are going to calculate the *real* maxUnavailable number when it is a percentage to avoid round down to `0`, the `replicas` will be affected. For example there was a StatefulSet with `replicas=10 and maxUnavailable=10%`, which is allowed. And then someone or HPA tries to change `replicas` to 9 and he/she/it will get error because the validation finds the `maxUnavailable = 9 * 10%` (with round down) will get `0` which is not allowed.

So we'd better guarantee the *real* maxUnavailable not to less than 1 when round down a percentage.

Reply:
Yeah, integer is fine for apis validation will check if it is positive.